daemon/logger/loggerutils/logfile.go

```diff
@@ -2,41 +2,107 @@ package loggerutils
 
 import (
 	"bytes"
+	"compress/gzip"
 	"context"
+	"encoding/json"
 	"fmt"
 	"io"
 	"os"
 	"strconv"
+	"strings"
 	"sync"
 	"time"
 
 	"github.com/docker/docker/daemon/logger"
 	"github.com/docker/docker/daemon/logger/loggerutils/multireader"
 	"github.com/docker/docker/pkg/filenotify"
+	"github.com/docker/docker/pkg/pools"
 	"github.com/docker/docker/pkg/pubsub"
 	"github.com/docker/docker/pkg/tailfile"
 	"github.com/fsnotify/fsnotify"
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 )
 
+const tmpLogfileSuffix = ".tmp"
+
+// RotateFileMetadata is a metadata of the gzip header of the compressed log file
+type RotateFileMetadata struct {
+	LastTime time.Time `json:"lastTime,omitempty"`
+}
+
+// RefCounter is a counter of logfile being referenced
+type RefCounter struct {
+	mu      sync.Mutex
+	counter map[string]int
+}
+
+// Reference increase the reference counter for specified logfile
+func (rc *RefCounter) Reference(fileName string, openRefFile func(fileName string) (*os.File, error)) (*os.File, error) {
```

Comment:
Maybe we can change this `openRefFile` to `func(fileName string, exists bool) (*os.File, error)`
Then the implementation here would be:
```go
_, ok := rc.counter[fileName]
f, err := open(fileName, ok)
```

It may make sense to also change the fn name to `GetReference` in that case.